mixing/vec.go

```diff
@@ -0,0 +1,79 @@
+package mixing
+
+import (
+	"bytes"
+	"fmt"
+	"strings"
+
+	"github.com/decred/dcrd/mixing/internal/chacha20prng"
+)
+
+const Msize = 20
+
+// Vec is a N-element vector of Msize []byte messages.
+type Vec [][Msize]byte
+
+func randVec(n uint32, prng *chacha20prng.Reader) Vec {
+	v := make(Vec, n)
+	for i := range v {
+		prng.Read(v[i][:])
+	}
+	return v
+}
+
+// Equals returns whether the two vectors have equal dimensions and data.
+func (v Vec) Equals(other Vec) bool {
+	if len(v) != len(other) {
+		return false
+	}
+	for i := range other {
+		if !bytes.Equal((v)[i][:], other[i][:]) {
+			return false
+		}
+	}
+	return true
+}
+
+// M returns the i'th message of the vector.
+// XXX: still necessary?
+// func (v Vec) M(i uint32) []byte {
+// 	return v[i][:]
+// }
+
+func (v Vec) String() string {
+	b := new(strings.Builder)
+	b.Grow(2 + len(v)*(2*Msize+1))
+	b.WriteString("[")
+	for i := range v {
+		if i != 0 {
+			b.WriteString(" ")
+		}
+		fmt.Fprintf(b, "%x", v[i][:])
```

Comment:
Hmm, that's something I'll have to test out.  I wasn't aware that would save a heap allocation.  I'm pretty sure that it's going to cause an allocation either way because it is forced to escape due to being an interface.

Reply:
Also for reference, it does in fact cause an alloc either way:

As written:
```sh
$ go build -gcflags="-m" 2>&1 | grep "vec.go:51.*v\[i\]"
./vec.go:51:28: v[i][:] escapes to heap
```

Without slicing:
```sh
$ go build -gcflags="-m" 2>&1 | grep "vec.go:51.*v\[i\]"
./vec.go:51:25: v[i] escapes to heap
```

Reply:
At any rate, I think you're probably right about it avoiding the copy though.  I'm curious to test that at some point, but it's fine as is.  Resolving.